{{ cookiecutter.repo_name }}/{{ cookiecutter.project_name }}/{{ cookiecutter.plugin_name.lower().replace(' ', '') }}plugin.py

```diff
@@ -10,56 +10,72 @@
+{%- if cookiecutter.graphical_plugin != 'n' %}
 from qtpy.QtWidgets import QVBoxLayout
-
-{% if cookiecutter.spyder3_compatibility == 'y' -%}
+{%- endif %}
+{% if cookiecutter.spyder3_compatibility == 'y' %}
 try:
     from spyder.api.plugins import SpyderPluginWidget
 except ImportError:
     from spyder.plugins import SpyderPluginWidget # Spyder3
     {%- if cookiecutter.create_config_page == 'y' %}
 try:
-    from spyder.api.preferences import
+    from spyder.api.preferences import PluginConfigPage
 except ImportError:
     from spyder.plugins.configdialog import PluginConfigPage # Spyder3
     {%- endif %}
-{% else -%}
+{% else %}
+    {%- if cookiecutter.graphical_plugin == 'y' %}
 from spyder.api.plugins import SpyderPluginWidget
+    {%- else %}
+from spyder.api.plugins import SpyderPlugin
+    {%- endif %}
     {%- if cookiecutter.create_config_page == 'y' %}
 from spyder.api.preferences import PluginConfigPage
-    {%- endif %}
-{%- endif -%}
-
+    {% endif %}
+{%- endif %}
 from .widgets.{{ cookiecutter.plugin_name.lower().replace(' ', '') }}gui import {{ widget_name }}
-{% if cookiecutter.create_config_page == 'y' %}
 
+{% if cookiecutter.create_config_page == 'y' %}
 class {{ config_page }}(PluginConfigPage):
     """{{ cookiecutter.plugin_name }} plugin preferences."""
     pass
 {% endif %}
-
+{%- if cookiecutter.graphical_plugin == 'n' %}
+class {{ cookiecutter.plugin_name.replace(' ', '') }}Plugin(SpyderPlugin):
+    """{{ cookiecutter.plugin_name }} plugin."""
+    CONF_SECTION = '{{ cookiecutter.plugin_name }}'
+    {% if cookiecutter.create_config_page == 'y' %}
+    CONFIGWIDGET_CLASS = {{ config_page }}
+    {% else %}
+    CONFIGWIDGET_CLASS = None
+    {% endif %}
+    def __init__(self, parent=None):
+        QObject.__init__(self, parent)
+        SpyderPlugin.__init__(self, parent)
+{%- else %}
 class {{ cookiecutter.plugin_name.replace(' ', '') }}Plugin(SpyderPluginWidget):
     """{{ cookiecutter.plugin_name }} plugin."""
     CONF_SECTION = '{{ cookiecutter.plugin_name }}'
-{% if cookiecutter.create_config_page == 'y' %}
+    {% if cookiecutter.create_config_page == 'y' %}
     CONFIGWIDGET_CLASS = {{ config_page }}
-{% else %}
+    {% else %}
     CONFIGWIDGET_CLASS = None
-{% endif %}
-
+    {% endif %}
     def __init__(self, parent=None):
+        QObject.__init__(self, parent)
         SpyderPluginWidget.__init__(self, parent)
-        self.main = parent # Spyder3
 
-        # Create widget and add to dockwindow
-        self.widget = {{ widget_name }}(self.main)
+        {% if cookiecutter.spyder3_compatibility == 'y' %}
+        # Initialize plugin
+        self.initialize_plugin()
+        {% endif %}
+
+        # Graphical view
         layout = QVBoxLayout()
         layout.addWidget(self.widget)
         self.setLayout(layout)
 
-        # Initialize plugin
-        self.initialize_plugin()
-
     # --- SpyderPluginWidget API ----------------------------------------------
     def get_plugin_title(self):
```

Comment:
The only mandatory methods for Spyder 4 are `get_plugin_title` and `on_first_registration`, but you can also leave here `get_focus_widget`.

Please add the others (i.e. refresh_plugin, get_plugin_actions, apply_plugin_settings, register_plugin) when the Spyder 3 compatibility is on.